rust/src/dataset/optimize.rs

```diff
@@ -0,0 +1,1015 @@
+    pub fragments: Vec<Fragment>,
+    pub pos_range: Range<usize>,
+    pub candidacy: Vec<CompactionCandidacy>,
+    pub row_counts: Vec<usize>,
+}
+
+impl CandidateBin {
+    /// Return true if compacting these fragments wouldn't do anything.
+    fn is_noop(&self) -> bool {
+        if self.fragments.is_empty() {
+            return true;
+        }
+        // If there's only one fragment, it's a noop if it's not CompactItself
+        if self.fragments.len() == 1 {
+            matches!(self.candidacy[0], CompactionCandidacy::CompactWithNeighbors)
+        } else {
+            false
+        }
+    }
+
+    /// Split into one or more bins with at least `min_num_rows` in them.
+    fn split_for_size(mut self, min_num_rows: usize) -> Vec<Self> {
+        let mut bins = Vec::new();
+
+        loop {
+            let mut bin_len = 0;
+            let mut bin_row_count = 0;
+            while bin_row_count < min_num_rows && bin_len < self.row_counts.len() {
+                bin_row_count += self.row_counts[bin_len];
+                bin_len += 1;
+            }
+
+            // If there's enough remaining to make another worthwhile bin, then
+            // push what we have as a bin.
+            if self.row_counts[bin_len..].iter().sum::<usize>() >= min_num_rows {
+                bins.push(Self {
+                    fragments: self.fragments.drain(0..bin_len).collect(),
+                    pos_range: self.pos_range.start..(self.pos_range.start + bin_len),
+                    candidacy: self.candidacy.drain(0..bin_len).collect(),
+                    row_counts: self.row_counts.drain(0..bin_len).collect(),
+                });
+                self.pos_range.start += bin_len;
+            } else {
+                // Otherwise, just push the remaining fragments into the last bin
+                bins.push(self);
+                break;
+            }
+        }
+
+        bins
+    }
+}
+
+/// Formulate a plan to compact the files in a dataset
+///
+/// The compaction plan will contain a list of tasks to execute. Each task
+/// will contain approximately `target_rows_per_fragment` rows and will be
+/// rewriting fragments that are adjacent in the dataset's fragment list. Some
+/// tasks may contain a single fragment when that fragment has deletions that
+/// are being materialized and doesn't have any neighbors that need to be
+/// compacted.
+pub async fn plan_compaction(
+    dataset: &Dataset,
+    options: &CompactionOptions,
+) -> Result<CompactionPlan> {
+    // We assume here that get_fragments is returning the fragments in a
+    // meaningful order that we want to preserve.
+    let mut fragment_metrics = futures::stream::iter(dataset.get_fragments())
+        .map(|fragment| async move {
+            match collect_metrics(&fragment).await {
+                Ok(metrics) => Ok((fragment.metadata, metrics)),
+                Err(e) => Err(e),
+            }
+        })
+        .buffered(num_cpus::get() * 2);
```

Comment:
Why times 2?

Reply:
I should leave a comment, but my intuition for "readahead" situations where it's waiting on IO we can use some multiple of the number of cpus. In contrast, for CPU bound tasks stick to `num_cpus::get()` without any multiplier. We should really consolidate this into some sort of configurable session...